### DISCUSSION TITLE
Why separate cipher-suite() configuration for TLS1.2 and TLS1.3?

### DISCUSSION THREAD

**Discussion by mohitvaid** [syslog-ng/syslog-ng#4007] [Q&A] [answered]
Hi Team, I was going through latest changes to be able to specify cipher-suites separately for TLS1.2 and TLS1.3.
My question is why have we come up with the following approach,

```
tls12-and-older() can be used to specify TLS v1.2-and-older ciphers,
tls13() can be used for TLS v1.3 ciphers only.
```
though just specifying:

```
cipher-suites() 
```

looks much more cleaner.

I am thinking in terms of scalability for future protocol support(a distant vision though). 

The challenge is, this per protocol specification puts burden at the configuration side, though I guess its better if the library abstracts out these details.

If we will follow the same design, then may be in future we will end up specifying, ecdh-curves-TLS1.2 and ecdh-curves-TLS1.3 etc.

**Comment by MrAnno** [comment, marked as answer]
Hi,

Thank you for the question.

OpenSSL decided to differentiate between TLS1.2-and-older and TLS 1.3 ciphers for various compatibility reasons. This is not the case with elliptic curves and most of the other options.

You can check the full discussion about this topic here: https://github.com/syslog-ng/syslog-ng/issues/3906#issuecomment-1033698573

In summary, Apache, nginx, OpenVPN, and others use separate options due to OpenSSL (so that they/we can avoid implementing our own hard-to-maintain lists to separate these two cipher categories). If OpenSSL decided to publish a unified config option (or API call), we'd be able to restore the intuitive behavior, but that would be a breaking change.

Note: As TLS configuration is getting more and more difficult, applications started supporting OpenSSL's SSL_CONF method to fine-tune anything TLS-related when using OpenSSL as TLS implementation. We're planning to do the same:

https://github.com/syslog-ng/syslog-ng/issues/3993

https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslopensslconfcmd

https://nginx.org/en/docs/http/ngx_http_ssl_module.html#ssl_conf_command
